Update DNS deploy logic and functions

johnwhitton opened the pull request:
This PR continues work from https://github.com/polymorpher/ens-deployer/pull/6

Some Notes

1. [dns-js problem with empty IP's](https://github.com/polymorpher/ens-deployer/pull/6#discussion_r1105393215)
Current workaround (before creating the PR and creating our own dns-js package) is to modify dnsrecord.js locally as follows
```
function writeA(out, ip) {
  if (ip === '') {return}
  var parts = ip.split('.');
  for (var i = 0; i < 4; i++) {
    out.byte(parts[i]);
  }
}
```
and add this line to lib/index.js
```
exports.BufferWriter = require('./bufferwriter');
```
2. When deploying DNS records for the TLD (`.country`) the node should be empty not the TLD
```
 const TLDnode = namehash.hash('')
  //   const TLDnode = namehash.hash(TLD)
  ```
 3. When encoding DNS Records we only need to return the hex String (not the EncodeARecord type) current fix until we decide if we have other reasons for wanting the additional information is
 **For now have updated to just return a single value and used objects for Record types see [here](https://github.com/polymorpher/ens-deployer/pull/7/commits/e62ace82b84c784c1999e0ae3be88ad0bbf16b9e)**
```
eexport const encodeARecord = (aRecord: ARecord): EncodedRecord['hexString'] => {
  // Sample Mapping
  // a.test.country. 3600 IN A 1.2.3.4
  /*
    name: a.test.country
    type: A
    class: IN
    ttl: 3600
    address: 1.2.3.4
  */
  // returns 0161047465737407636f756e747279000001000100000e10000401020304

  // a empty address is used to remove existing records
  const EncodedRecord = encodeRecord({
    name: aRecord.name,
    type: DNSRecord.Type.A,
    class: DNSRecord.Class.IN,
    ttl: 3600,
    address: aRecord.ipAddress
  })
  return EncodedRecord.hexString
}
```

johnwhitton committed:
Update DNS deploy logic and functions

johnwhitton committed:
remove redundant code

johnwhitton committed:
Update Type Definitions for encoding records

polymorpher commented:
We need to fork dns-js and fix the bug. See comment in #6 and modified README. We should not ask people or to rely on modifying local packages in node_modules because they will be overwritten each time yarn does something related to all packages.

We need to return both full DNS record and the buffer for debugging and comparison purposes (e.g. `debugPrintRecord`). Hex string is returned in position 0 because it results in the least amount of code modification / for convenience

johnwhitton committed:
change type aliases to interface for DNS Records

polymorpher commented:
If it is a DNS record for the root node, the variable should be named root node and the code should make it clear. It is also unclear why this record is set at root level. What is the general rule for DNS record lookup, what zones are looked up, and how exactly does a specific implementation (e.g. the 1ns plugin) does it?

polymorpher commented on contract/lib/dns.ts:
```diff
@@ -1,17 +1,6 @@
-}
-// TODO: define members of object for each type of record
-type EncodedRecord = [string, Buffer, object]
+import type { EncodedRecord, ARecord, CNAMERecord, SOARecord, TXTRecord } from './types.d.ts'
```
drop .ts. It should also be just "import"

johnwhitton replied:
Done :pray:

polymorpher commented on contract/lib/dns.ts:
```diff
@@ -44,11 +33,11 @@ export const debugPrintRecord = (record: object, bufferWriter: Buffer) => {
 }
 
 // TODO: buggy for empty ipAddress. Must fix later
-export const encodeARecord = (name: string, ipAddress: string): EncodedRecord => {
```
The value is a single type, there is no reason to wrap it in an object. Same as others. I already added whatever types that were necessary

johnwhitton replied:
Reverted 🙏

johnwhitton committed:
Replace dns-js with 1ns-js to handle null IP addresses

johnwhitton commented:
> We need to fork dns-js and fix the bug. See comment in #6 and modified README. We should not ask people or to rely on modifying local packages in node_modules because they will be overwritten each time yarn does something related to all packages.

Have created package [1ns-dns-js](https://www.npmjs.com/package/1ns-dns-js) and updated to use this in [this commit](https://github.com/polymorpher/ens-deployer/pull/7/commits/4fe7a4cbd755c49c2a8b3ce3b2365fd7a23b3aa9)

johnwhitton committed:
revert EncodedRecord back to an array, remove erroneous DNSRecords from TLD

johnwhitton commented:
> We need to return both full DNS record and the buffer for debugging and comparison purposes (e.g. `debugPrintRecord`). Hex string is returned in position 0 because it results in the least amount of code modification / for convenience

My apologies was confused by the elegance of the solution :)
I have reverted EncodeRecord back to an array and updated the calls to use the first element. [see here](https://github.com/polymorpher/ens-deployer/pull/7/commits/ad5819b795f69e9fbd9f4aa039437d5923713ca5)

johnwhitton commented:
> If it is a DNS record for the root node, the variable should be named root node and the code should make it clear. It is also unclear why this record is set at root level. What is the general rule for DNS record lookup, what zones are looked up, and how exactly does a specific implementation (e.g. the 1ns plugin) does it?

Have removed for now. I don't believe a records should be set for the TLD. (commit is [here](https://github.com/polymorpher/ens-deployer/pull/7/commits/ad5819b795f69e9fbd9f4aa039437d5923713ca5))

🙏

johnwhitton committed:
Include support for DNAME, NS and SRV records

johnwhitton committed:
Remove sample zonefile header from test README.md

johnwhitton committed:
Update testing for all record types

polymorpher commented:
Can you resolve the merge conflicts

johnwhitton committed:
Merge upstream/main

johnwhitton commented:
> Can you resolve the merge conflicts

Done 🙏

johnwhitton commented:
If you want to merge https://github.com/polymorpher/node-dns-js/pull/1
I'll update package.json to use https://github.com/polymorpher/node-dns-js
In the mean time I'm using 
`        "dns-js": "git+https://github.com/jw-1ns/node-dns-js",`

johnwhitton committed:
use polymorpher node-dns-js

johnwhitton commented:
Update should be ready to merge 🙏

polymorpher commented on contract/yarn.lock:
```diff
@@ -3862,9 +3862,9 @@ dir-glob@^3.0.1:
     path-type "^4.0.0"
 
-"dns-js@git+https://github.com/polymorpher/node-dns-js":
+"dns-js@git+https://github.com/jw-1ns/node-dns-js":
```
need to run yarn install again

polymorpher commented on yarn.lock:
```diff
@@ -0,0 +1,4 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
should remove file

johnwhitton committed:
use polymorpher node-dns-js this time with yarn install
